core/src/extension.rs

```diff
@@ -0,0 +1,40 @@
+use crate::plugin::Plugin;
+use crate::uri::{Uri, UriBound};
+use std::any::Any;
+use std::marker::PhantomData;
+use std::os::raw::c_void;
+
+pub unsafe trait Extension<P: Plugin>: UriBound {
+    const RAW_DATA: &'static Any;
+
+    const DESCRIPTOR: ExtensionDescriptor<P> = ExtensionDescriptor {
+        uri: Self::URI,
+        raw_data: Self::RAW_DATA as *const _ as *mut _,
+        _plugin: PhantomData,
+    };
+}
+
+pub struct ExtensionDescriptor<P: Plugin> {
```

Comment:
It is here so an extension descriptor is virtually "tied" to a specific plugin implementation, so a plugin may not "steal" an extension descriptor from another plugin in the same library (without using unsafe code of course).

Here is an example of what could happen without this type argument:

```rust
struct FooPlugin;
struct BarPlugin;

impl lv2_bar::BarExtension for FooPlugin {
    fn bar(&self) -> i32 {
        69
    }
}


impl Plugin for BarPlugin {
    const EXTENSIONS: &'static [lv2_core::extension::ExtensionDescriptor] = &[
        <lv2_bar::BarExtension as lv2_core::extension::Extension<FooPlugin>>::DESCRIPTOR, // Uh-oh
    ];

    /* [...] */
}
```

This code would compile here, since for the compiler, `ExtensionDescriptor`s for both `FooPlugin` and `BarPlugin` are the same type, so it will happily let you swap them this way.
This would cause the host to call the `bar` implementation for `FooPlugin` with an instance of `BarPlugin` instead, and here UB ensues.

However, with this additional type parameter, the compiler prevents it:

```
error[E0308]: mismatched types
   --> core/tests/extension.rs:112:9
    |
112 |         <lv2_bar::BarExtension as lv2_core::extension::Extension<FooPlugin>>::DESCRIPTOR, // Uh-oh
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `TestPlugin`, found struct `FooPlugin`
    |
    = note: expected type `lv2_core::extension::ExtensionDescriptor<TestPlugin>`
               found type `lv2_core::extension::ExtensionDescriptor<FooPlugin>`
```

I know this soundness hole is a bit far-fetched, but for the end-user it is just a matter of adding `<Self>`, and if they use the `lv2_extensions!` helper macro they don't even see it, so in the end I think it is better to have it in. :slightly_smiling_face:

Reply:
Oh, that's cool! I like it!